src/memfs.ts

```diff
@@ -0,0 +1,288 @@
+			return content;
+		}
+
+		return getTypeScriptLibraries().get(path);
+	}
+
+	/**
+	 * Invalidates temporary content denoted by the given path
+	 * @param path path to a file relative to project root
+	 */
+	didClose(path: string) {
+		this.overlay.delete(path);
+	}
+
+	/**
+	 * Adds temporary content denoted by the given path
+	 * @param path path to a file relative to project root
+	 */
+	didSave(path: string) {
+		const content = this.readFileIfExists(path);
+		if (content !== undefined) {
+			this.add(util.path2uri('', path), content);
+		}
+	}
+
+	/**
+	 * Updates temporary content denoted by the given path
+	 * @param path path to a file relative to project root
+	 */
+	didChange(path: string, text: string) {
+		this.overlay.set(path, text);
+	}
+
+	/**
+	 * Called by TS service to scan virtual directory when TS service looks for source files that belong to a project
+	 */
+	readDirectory(rootDir: string, extensions: string[], excludes: string[], includes: string[]): string[] {
+		return match.matchFiles(rootDir,
+			extensions,
+			excludes,
+			includes,
+			true,
+			this.path,
+			p => this.getFileSystemEntries(p));
+	}
+
+	/**
+	 * Called by TS service to scan virtual directory when TS service looks for source files that belong to a project
+	 */
+	getFileSystemEntries(path: string): match.FileSystemEntries {
+		const ret: { files: string[], directories: string[] } = { files: [], directories: [] };
+		let node = this.rootNode;
+		const components = path.split('/').filter(c => c);
+		if (components.length !== 1 || components[0]) {
+			for (const component of components) {
+				const n = node.children.get(component);
+				if (!n) {
+					return ret;
+				}
+				node = n;
+			}
+		}
+		node.children.forEach((value, name) => {
+			if (value.file) {
+				ret.files.push(name);
+			} else {
+				ret.directories.push(name);
+			}
+		});
+		return ret;
+	}
+
+	trace(message: string) {
+		this.logger.log(message);
+	}
+}
+
+/**
+ * Iterates over in-memory cache calling given function on each node until callback signals abort or all nodes were traversed
+ */
+export function walkInMemoryFs(fs: InMemoryFileSystem, rootdir: string, walkfn: (path: string, isdir: boolean) => Error | void): Error | void {
+	const err = walkfn(rootdir, true);
+	if (err) {
+		if (err === skipDir) {
+			return;
+		}
+		return err;
+	}
+	const { files, directories } = fs.getFileSystemEntries(rootdir);
+	for (const file of files) {
+		const err = walkfn(path_.posix.join(rootdir, file), false);
+		if (err) {
+			return err;
+		}
+	}
+	for (const dir of directories) {
+		const err = walkInMemoryFs(fs, path_.posix.join(rootdir, dir), walkfn);
+		if (err) {
+			return err;
+		}
+	}
+	return;
+}
+
+/**
+ * TypeScript library files fetched from the local file system (bundled TS)
+ */
+let tsLibraries: Map<string, string>;
+
+/**
+ * Fetches TypeScript library files from local file system
+ */
+export function getTypeScriptLibraries(): Map<string, string> {
```

Comment:
Instead of this getter function, lets just use the map directly. Does it even have to be exported?